pkgs/pkgs-lib/formats.nix

```diff
@@ -58,9 +59,121 @@ rec {
+    arbitrary secret values.
+
+    Example:
+      If the file "/path/to/secret" contains the string
+      "topsecretpassword1234",
+
+      genJqSecretsReplacementSnippet
+        {
+          example = [
+            {
+              irrelevant = "not interesting";
+            }
+            {
+              ignored = "ignored attr";
+              relevant = {
+                secret = lib.mkSecret "/path/to/secret";
+              };
+            }
+          ];
+        }
+        "/path/to/output.json"
+
+      would generate a snippet that, when run, outputs the following
+      JSON file at "/path/to/output.json":
+
+      {
+        "example": [
+          {
+            "irrelevant": "not interesting"
+          },
+          {
+            "ignored": "ignored attr",
+            "relevant": {
+              "secret": "topsecretpassword1234"
+            }
+          }
+        ]
+      }   */
+    genJqSecretsReplacementSnippet = set: output:
+      let /*
+        Recurse into a list or an attrset, searching for secrets, and
+        return an attrset where the names are the corresponding jq
+        path where the attrs were found and the values are the values
+        of the attrs.
+
+        Example:
+          getSecretsWithJqPrefix
+            {
+              example = [
+                {
+                  irrelevant = "not interesting";
+                }
+                {
+                  ignored = "ignored attr";
+                  relevant = {
+                    secret = lib.mkSecret "/path/to/secret";
+                  };
+                }
+              ];
+            }
+            -> { ".example[1].relevant.secret" = "/path/to/secret"; }   */
+        getSecretsWithJqPrefix = item:
+          let
+            recurse = prefix: item:
+              if lib.isSecret item then
+                lib.nameValuePair prefix item._secret
+              else if builtins.isAttrs item then
+                lib.mapAttrsToList (name: value: recurse (prefix + "." + name) value) item
+              else if builtins.isList item then
+                lib.imap0 (index: item: recurse (prefix + "[${toString index}]") item) item
+              else
+                [];
+          in lib.listToAttrs (lib.flatten (recurse "" item));
+
+        secrets = getSecretsWithJqPrefix set;
+        setFile = pkgs.writeText output (builtins.toJSON set);
+      in ''
+        ${lib.concatStringsSep "\n"
+            (lib.imap1
+              (index: name: ''
+                 secret${toString index}=$(<${lib.escapeShellArg secrets.${name}})
+                 export secret${toString index}
+               '')
+              (builtins.attrNames secrets))}
+        ${pkgs.jq}/bin/jq \
+          <${setFile} \
+          >${lib.escapeShellArg output} \
+          '${lib.concatStringsSep " | "
+               (lib.imap1
+                  (index: name: "${name} = $ENV.secret${toString index}")
+                  (builtins.attrNames secrets))}'
```

Comment:
This probably has escaping issues. Try to pass all variables with command line arguments like `--arg` so that the jq script itself is static and doesn't need Nix interpolation

Reply:
This is an example of how a generated script looks:

```bash
set -o errexit -o pipefail -o nounset -o errtrace
shopt -s inherit_errexit

umask u=rwx,g=,o=
secret1=$(<'/run/keys/aws_access_key')
export secret1

secret2=$(<'/run/keys/aws_access_key')
export secret2

secret3=$(<'/run/keys/oidc_secret')
export secret3

/nix/store/q4rjxagxikpsbvir0clyxk24rn2wckgg-jq-1.6-bin/bin/jq \
  </nix/store/s5nldkzagsydw7pnxqzhc8qwn4iw7dka--var-gitlab-state-config-gitlab.yml-tmp \
  >'/var/gitlab/state/config/gitlab.yml-tmp' \
  '.production.backup.upload.connection.aws_secret_access_key = $ENV.secret1 | .production.object_store.connection.aws_secret_access_key = $ENV.secret2 | .production.omniauth.providers[0].args.client_options.secret = $ENV.secret3'
```

To me, there aren't any obvious places where escaping issues would sneak in, except maybe where the file contents are read into the indexed variables. An alternative could be to use `--slurpfile`, but I have a vague memory of trying that and it not working as expected.

Reply:
Escaping problems come from the attribute names, because those attributes need to be escaped both for jq and shell. This means you can easily run arbitrary code from an attribute name, like this: `{ "'$(launch_nukes)'" = mkSecret "/some/path"; }`. Problems like this are avoided nicely by using `--arg` in `jq` and to try to avoid concatenating arbitrary strings in general

Reply:
Oh, you're referring to the json paths.. I don't really see a way to get rid of those here. `--arg` only sets variables, the values of which would still have to be read in a filter which names which json path they should be applied to.

Reply:
Sure, but we could just put the file path straight into the exported json we're transforming in that case - the additional json file would just add a layer of indirection. The problem is that we don't, to my knowledge, have a way to do the substitution in jq, since you can't read arbitrary raw files at runtime in filters.

Reply:
Maybe it would be easier to just go the ini route and do a text replacement on the sha256 of the path? I just had this implemented for json already (it's from `nixos/lib/utils.nix`), so I figured it would be easier to reuse it.